PcAtChipsetPkg/PcatRealTimeClockRuntimeDxe/PcRtc.c

```diff
@@ -324,21 +326,61 @@ PcRtcInit (
     Time.Nanosecond = 0;
     Time.TimeZone   = EFI_UNSPECIFIED_TIMEZONE;
     Time.Daylight   = 0;
+    NeedRtcUpdate = TRUE;
   }
 
   //
   // Set RTC configuration after get original time
   // The value of bit AIE should be reserved.
   //
-  RegisterB.Data = FixedPcdGet8 (PcdInitialValueRtcRegisterB) | (RegisterB.Data & BIT5);
-  RtcWrite (RTC_ADDRESS_REGISTER_B, RegisterB.Data);
+  if ((RegisterB.Data | BIT5) != (FixedPcdGet8 (PcdInitialValueRtcRegisterB) | BIT5)) {
+    RegisterB.Data = FixedPcdGet8 (PcdInitialValueRtcRegisterB) | (RegisterB.Data & BIT5);
+    RtcWrite (RTC_ADDRESS_REGISTER_B, RegisterB.Data);
+    NeedRtcUpdate = TRUE;
+  }
 
   //
   // Reset time value according to new RTC configuration
   //
-  Status = PcRtcSetTime (&Time, Global);
-  if (EFI_ERROR (Status)) {
-    return EFI_DEVICE_ERROR;
+  if (NeedRtcUpdate) {
+    Status = PcRtcSetTime (&Time, Global);
+    if (EFI_ERROR (Status)) {
+      return EFI_DEVICE_ERROR;
+    }
+  }
+  else {
+    //
+    // Write timezone and daylight to RTC variable
+    //
+    if ((Time.TimeZone == EFI_UNSPECIFIED_TIMEZONE) && (Time.Daylight == 0)) {
+      Status = EfiSetVariable (
+                 mTimeZoneVariableName,
+                 &gEfiCallerIdGuid,
+                 0,
+                 0,
+                 NULL
+                 );
+      if (Status == EFI_NOT_FOUND) {
+        Status = EFI_SUCCESS;
+      }
+    } else {
+      TimerVarNew = Time.Daylight;
+      TimerVarNew = (UINT32)((TimerVarNew << 16) | (UINT16)(Time.TimeZone));
+      if (TimerVarNew != TimerVar) {
+        Status   = EfiSetVariable (
+                     mTimeZoneVariableName,
+                     &gEfiCallerIdGuid,
+                     EFI_VARIABLE_BOOTSERVICE_ACCESS | EFI_VARIABLE_RUNTIME_ACCESS | EFI_VARIABLE_NON_VOLATILE,
+                     sizeof (TimerVarNew),
+                     &TimerVarNew
+                     );
+      }
+    }
+    if (EFI_ERROR (Status)) {
+      return EFI_DEVICE_ERROR;
+    }
+    Global->SavedTimeZone = Time.TimeZone;
+    Global->Daylight      = Time.Daylight;

```

Comment:
How about separate the code in a function and then call it from here and PcRtcSetTime()?